### PR TITLE
[Codegen][GPU] Add pass to reuse shared memory buffers in simple cases

### PR DESCRIPTION
**Pull request by Max191**
This PR adds a new pass that tries to reuse shared memory allocations in functions. This pass only does a very basic analysis, assuming no control flow operations (and is conservative with structured control flow), but it is useful for some of the cases that show up in practice with the current pipelines.

### CODE REVIEW

**Comment by Max191** [comment]
Not adding the pass anywhere yet, since I'm not sure what the best place to put it is yet. Going to wait on @hanhanW's work for TK kernels.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPUUtils.cpp
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPUUtils.cpp]: Oops, I need to make this an option, sorry

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp:43]: nit: no auto here, the type is not obvious

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]:
```suggestion
     Operation* op = workList.pop_back_val();
```

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]:
```suggestion
  return {begin, end};
```

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]: nit: no auto here

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]: nit: no auto

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]: nit: no auto

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]: nit: no auto

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]: nit: no auto

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]: nit: inline this

**Review by kuhar — COMMENTED** [review]
Do we need a test with `select` of two memrefs and/or subviews? Not sure if we have that or if the only way to express it is through `scf.if`

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]: `Codegen/Common/GPU` cannot depend on LLVMGPU. I would move the relevant utility to `GPU/Utils.h`

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]: add `isa<memref::SubViewOp, memref::CollapseShapeOp, memref::ExpandShapeOp>` too as those also have view-like behavior.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]: Same here, add collapse and expand.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]:
```suggestion
/// Sink all shared memory allocations and their subviews within the CFG, so
```

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]: Why are you doing this movement? Shared memory is a property of the kernel, so the location of the `memref.alloc` does not matter at all later on.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp:120]: Add a short description/example of the analysis done here. In particular highlight that this algorithm will greedily alias groups together in cases such as this:
```
%0 = alloc
%1 = alloc
%2 = alloc
%rd0 = read %0
%rd1 = read %1
%w0 = write %0
%rd2 = read %2
%w1 = write %1
%w2 = write %2
```
where alloc `%0` and `%2` are non-overlapping, but because `%1` overlaps with both of them, all three allocations will stay separate.

**Review by qedawkins — COMMENTED** [review]
LGTM after a few minor comments (and addressing Jakub's comments)

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (Max191) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]: The type name is quite long since it is a DenseMap pair, and the types of the `first` and `second` element of the pair are shown right below. Is it okay to keep this one auto?

Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]: Maybe use structured bindings instead `auto [key, val] : livenessMaps`, except with better names

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (Max191) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]: This was because of the util I was using to place barriers. I rewrote it without the util so I don't need this anymore. It also removes the LLVMGPU dependency.

File: compiler/src/iree/compiler/Codegen/Common/GPU/test/BUILD.bazel
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/test/BUILD.bazel]: Please fix the filename typo before landing
```suggestion
            "gpu_reuse_shared_memory_allocs.mlir",
```

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp]: Put this in an `LLVM_DEBUG`?

**Review by qedawkins — APPROVED** [review]

**Pull request merged**
